src/pyro/source.clj

```diff
@@ -23,13 +25,41 @@
   [s n]
   (str "--> " (pad-integer n) " " s))
 
+(defn filepath->lastModified
+  [filepath]
+  (let [url (io/resource filepath)
+        file (io/file filepath)]
+    (cond
+      ; handle files inside JARs
+      ; (io/resource "clojure/java/io.clj") =>
+      ; URL "jar:file:/Users/ivref/.m2/repository/org/clojure/clojure/1.8.0/clojure-1.8.0.jar!/clojure/java/io.clj"
+      (and url (= "jar" (.getProtocol url))) (-> url
+                                                 (.getFile)
+                                                 (str/split #"!")
+                                                 (first)
+                                                 (subs 5) ; remove file:
+                                                 (io/file)
+                                                 (.lastModified))
+
+      ; handle local files loaded using require and similar
+      ; (io/resource "pyro/source.clj") =>
+      ; URL "file:/Users/ivref/clojure/pyro/src/pyro/source.clj"
+      (and url (= "file" (.getProtocol url))) (-> url (.getFile) (io/file) (.lastModified))
+
+      ; handle files loaded using Cursive, load-file and similar
+      (.exists file) (.lastModified file)
+
+      :else (throw (ex-info "Could not find lastModified of filepath" {:filepath filepath})))))
+
 (defn filepath->buffered-reader
   [filepath]
-  (when-let [strm (.getResourceAsStream (RT/baseLoader) filepath)]
-    (BufferedReader. (InputStreamReader. strm))))
+  (if-let [strm (or (.getResourceAsStream (RT/baseLoader) filepath)
+                    (io/input-stream (io/file filepath)))]
+    (BufferedReader. (InputStreamReader. strm StandardCharsets/UTF_8))
+    (throw (ex-info "Could not get stream" {:filepath filepath}))))
```

Comment:
Again, I have a distaste for throwing exceptions in this library. We should assume nils are a reasonable return value, which means sticking with `when-let` and handling nils upstream.